### PR TITLE
Support creating output directories that only contain symlinks

### PR DESCRIPTION
**Pull request by jakemac53**
Closes https://github.com/dart-lang/build/issues/1242

Added a new `--symlink` option, which is defaulted to true for `test` but no other commands.

This is unfortunately a breaking change in build_runner and build_runner_core as it required adding a new interface which must be implemented by any `RunnerAssetReader`, which is an exposed class from both of these packages.

### CODE REVIEW

**Comment by jakemac53** [comment]
In terms of performance, when testing with the `pub` package and the new vm compilation pipeline it takes creating the merged directory from >40 seconds to <1 second. Note that the output directory is unnecessarily large since we don't clean up the modular dill files in that case, or the original dart files, but its still a good proxy I think for how this can affect large packages.

**Comment by jakemac53** [comment]
Tried running this with angulars _tests and looks like web tests don't work :man_facepalming: . Investigating now... but will require a fix in package:test it looks like.

File: build_runner/CHANGELOG.md [hunk 1]
```diff
@@ -1,3 +1,23 @@
+## 0.10.0-dev
+
+### Breaking Changes
+
+- The `RunnerAssetReader` interface now requires that you implement the new
```
[the remark this answers is not in the document]
Comment (jakemac53) [reply build_runner/CHANGELOG.md]: done

**Comment by jakemac53** [comment]
cc @natebosch ok I updated this with the changes we talked about.

There is now a `finalizeBuild` method on `BuildEnvironment`, and we use that to create the merged output dir. Only the `IOEnvironment` knows about the new asset reader interface.

**Comment by jakemac53** [comment]
Looks like symlinks don't work as expected on windows even though there is supposedly support for linking individual files according to windows docs, the VM always ends up creating a directory symlink.

**Comment by jakemac53** [comment]
Filed https://github.com/dart-lang/sdk/issues/33966

File: build_runner_core/lib/src/environment/io_environment.dart [hunk 2]
```diff
@@ -17,6 +17,8 @@ import '../package_graph/package_graph.dart';
 import 'build_environment.dart';
 import 'create_merged_dir.dart';
 
+final _logger = new Logger('IOEnvironment');
```
Comment (natebosch) [inline build_runner_core/lib/src/environment/io_environment.dart]: need to include this at https://github.com/dart-lang/build/blob/a10c67f7ef87bae6e942cd8ccc7dfc1903e5afa1/build_runner/lib/src/logging/std_io_logging.dart#L69

Comment (jakemac53) [reply build_runner_core/lib/src/environment/io_environment.dart]: done

**Pull request merged**